types/wlr_xdg_foreign_v2.c

```diff
@@ -0,0 +1,411 @@
+	wl_list_for_each(child, &imported->children, link) {
+		if (child->surface == wlr_surface_child) {
+			return;
+		}
+	}
+
+	child = calloc(1, sizeof(struct wlr_xdg_imported_child_v2));
+	if (child == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+	child->surface = wlr_surface_child;
+	child->xdg_surface_unmap.notify = handle_child_xdg_surface_unmap;
+	child->xdg_toplevel_set_parent.notify = handle_xdg_toplevel_set_parent;
+
+	if (wlr_surface_is_xdg_surface(wlr_surface)) {
+		struct wlr_xdg_surface *surface =
+			wlr_xdg_surface_from_wlr_surface(wlr_surface);
+		struct wlr_xdg_surface *surface_child =
+			wlr_xdg_surface_from_wlr_surface(wlr_surface_child);
+
+		wlr_xdg_toplevel_set_parent(surface_child, surface);
+		wl_signal_add(&surface_child->events.unmap,
+				&child->xdg_surface_unmap);
+		wl_signal_add(&surface_child->toplevel->events.set_parent,
+				&child->xdg_toplevel_set_parent);
+	}
+
+	wl_list_insert(&imported->children, &child->link);
+}
+
+static const struct zxdg_imported_v2_interface xdg_imported_impl = {
+	.destroy = xdg_imported_handle_destroy,
+	.set_parent_of = xdg_imported_handle_set_parent_of
+};
+
+static struct wlr_xdg_exported_v2 *xdg_exported_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zxdg_exported_v2_interface,
+		&xdg_exported_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void xdg_exported_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static const struct zxdg_exported_v2_interface xdg_exported_impl = {
+	.destroy = xdg_exported_handle_destroy
+};
+
+static void xdg_exporter_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct wlr_xdg_foreign_v2 *xdg_foreign_from_exporter_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zxdg_exporter_v2_interface,
+		&xdg_exporter_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void disconnect_imported(struct wlr_xdg_imported_v2 *imported) {
+	if (imported->exported != NULL) {
+		imported->exported = NULL;
+		zxdg_imported_v2_send_destroyed(imported->resource);
+
+		wl_list_remove(&imported->exported_destroyed.link);
+		wl_list_init(&imported->exported_destroyed.link);
+	}
+}
+
+static void xdg_exported_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_xdg_exported_v2 *exported =
+		xdg_exported_from_resource(resource);
+
+	wlr_signal_emit_safe(&exported->base.events.destroy, NULL);
+	wlr_xdg_foreign_exported_finish(&exported->base);
+
+	wl_list_remove(&exported->xdg_surface_unmap.link);
+	wl_list_remove(&exported->link);
+	free(exported);
+}
+
+static void handle_xdg_surface_unmap(
+		struct wl_listener *listener, void *data) {
+	struct wlr_xdg_exported_v2 *exported =
+		wl_container_of(listener, exported, xdg_surface_unmap);
+	wl_resource_destroy(exported->resource);
```

Comment:
This destroys the resource behind the client's back. We can't do this, otherwise the client will die when sending an `xdg_exported.destroy` request.

Why do we need to handle `unmap` in the first place? The protocol doesn't seem to mention anything about it.

Reply:
A good question. I will remove that as well.

Reply:
Hmm, I suppose we want to listen not for unmap, but for the case where the toplevel is destroyed.